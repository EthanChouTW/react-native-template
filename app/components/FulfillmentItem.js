import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { TouchableOpacity, StyleSheet, Text, View } from 'react-native';
import Locales, { currencyPrice } from '../locales';
import Theme from '../utils/styleCollection';

const styles = StyleSheet.create({
  row: {
    paddingBottom: 8,
    paddingTop: 8,
    paddingLeft: 16,
    paddingRight: 16,
    marginTop: -1,
    backgroundColor: 'white',
    flexDirection: 'column',
    alignItems: 'center',
    justifyContent: 'space-between'
  },
  itemContainer: {
    flexDirection: 'row'
  },
  quantityContainer: {
    flex: 1
  },
  titleContainer: {
    flex: 5
  },
  amountContainer: {
    flex: 2,
    alignItems: 'flex-end'
  },
  addonContainer: {
    marginTop: 5
  },
  quantity: {
    color: 'black',
    fontWeight: 'bold',
    fontSize: 16
  },
  quantityMoreThanOne: {
    color: Theme.hbRed,
    fontWeight: 'bold',
    fontSize: 16
  },
  title: {
    color: 'black',
    fontSize: 16
  },
  amount: {
    color: 'black',
    fontSize: 16
  },
  editButton: {
    marginTop: 5,
    flex: 1,
    alignItems: 'flex-start'
  },
  editButtonTitle: {
    fontWeight: 'bold',
    color: Theme.algaeGreenColor
  },
  addonItem: {
    color: Theme.hbLightGray
  },
  notes: {
    paddingTop: 5,
    color: Theme.hbRed,
    fontSize: 14
  }
});

export default class FulfillmentItem extends Component {
  static propTypes = {
    item: PropTypes.shape({
      itemTitle: PropTypes.string,
      quantity: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      amount: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
      currency: PropTypes.string
    }).isRequired,
    fulfillmentId: PropTypes.number,
    onPressEditButton: PropTypes.func,
    partialFulfillmentEnabled: PropTypes.bool
  };

  render() {
    const { fulfillmentId, item, onPressEditButton, partialFulfillmentEnabled } = this.props;
    const renderAddonItems = () =>
    (
      item.additionalSetItems
        ? Object.keys(item.additionalSetItems).map(addonKey => {
          const quantityDisplay =
              item.additionalSetItems[addonKey].quantity > 1
                ? `(x${item.additionalSetItems[addonKey].quantity})`
                : '';
          const titleDisplay = item.additionalSetItems[addonKey].title;
          return (
            <Text key={addonKey} style={styles.addonItem}>
              {`${titleDisplay} ${quantityDisplay}`}
            </Text>
          );
        })
        : null);

    const renderItemNotes = () =>
    (item.notes ? (
      <Text style={styles.notes}>
        {`${item.notes}`}
      </Text>
      ) : null);

    const renderEditQuantity = () => {
      if (partialFulfillmentEnabled) {
        return (
          <TouchableOpacity
            style={styles.editButton}
            visible={partialFulfillmentEnabled}
            onPress={() => {
              onPressEditButton(fulfillmentId, item);
            }}
          >
            <Text style={styles.editButtonTitle}>{Locales.t('lbl_edit').toUpperCase()}</Text>
          </TouchableOpacity>
        );
      }
      return null;
    };

    const itemQuantity = item.quantityInCart !== undefined ? item.quantityInCart : item.quantity;

    return (
      <View style={styles.row}>
        <View style={styles.itemContainer}>
          <View style={styles.quantityContainer}>
            <Text
              style={this.props.item.quantity > 1 ? styles.quantityMoreThanOne : styles.quantity}
            >
              x{parseInt(itemQuantity, 10)}
            </Text>
            {renderEditQuantity()}
          </View>
          <View style={styles.titleContainer}>
            <Text style={styles.title}>{this.props.item.itemTitle}</Text>
            <View style={styles.addonContainer}>{renderAddonItems()}</View>
            {renderItemNotes()}
          </View>
          <View style={styles.amountContainer}>
            <Text style={styles.amount}>
              {currencyPrice(this.props.item.amount, this.props.item.currency)}
            </Text>
          </View>
        </View>
      </View>
    );
  }
}
