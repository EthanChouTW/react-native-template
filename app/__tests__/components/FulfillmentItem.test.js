import 'react-native';
import React from 'react';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import FulfillmentItem from '../../components/FulfillmentItem';

describe('FulfillmentItem', () => {
  const item = { quantity: '1', itemTitle: 'title', amount: '100' };
  test('renders correctly', () => {
    const snapshot = renderer.create(<FulfillmentItem item={item} />).toJSON();
    expect(snapshot).toMatchSnapshot();
  });
});
