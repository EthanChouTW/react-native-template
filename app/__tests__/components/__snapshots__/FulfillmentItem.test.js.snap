// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FulfillmentItem renders correctly 1`] = `
<
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "white",
      "flexDirection": "column",
      "justifyContent": "space-between",
      "marginTop": -1,
      "paddingBottom": 8,
      "paddingLeft": 16,
      "paddingRight": 16,
      "paddingTop": 8,
    }
  }
>
  <
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        x
        1
      </>
    </>
    <
      style={
        Object {
          "flex": 5,
        }
      }
    >
      <
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 16,
          }
        }
      >
        title
      </>
      <
        style={
          Object {
            "marginTop": 5,
          }
        }
      />
    </>
    <
      style={
        Object {
          "alignItems": "flex-end",
          "flex": 2,
        }
      }
    >
      <
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "black",
            "fontSize": 16,
          }
        }
      >
        100
      </>
    </>
  </>
</>
`;
